batch/batch/batch.py

```diff
@@ -998,10 +998,58 @@ def is_successful(self):
 
 
 async def _cancel_batch(batch_id, user):
-    batch = await Batch.from_db(batch_id, user)
-    if not batch:
-        abort(404)
-    await batch.cancel()
+    async with db.pool.acquire() as conn:
+        async with conn.cursor() as cursor:
+            await cursor.execute('''
+    update jobs
+inner join batch on jobs.batch_id == batch.batch_id
+       set jobs.state = 'Cancelled'
+     where jobs.batch_id = %s
+       and jobs.user = %s
+       and batch.deleted = FALSE
+       and batch.cancelled = FALSE
+       and jobs.state not in ('Cancelled', 'Error', 'Failed', 'Success', 'Pending')
+       and jobs.always_run == FALSE
+''', (batch_id, user))
+            await cursor.execute('''
+    update jobs
+inner join (    select jobs.id
+                  from jobs
+            inner join `jobs-parents` on `jobs-parents`.batch_id = jobs.batch_id
+                                     and `jobs-parents`.job_id = jobs.job_id
+                 where jobs.state = 'Pending'
+                   and jobs.batch_id = %s
+                   and jobs.user = %s
+                   and batch.deleted = FALSE
+                   and batch.cancelled = FALSE
+              group by jobs.job_id
+                having count(`jobs-parents`.state = 'Success') = count(*)
+                    or (jobs.always_run = TRUE and
+                        count(`jobs.parents`.state in ('Success', 'Failed', 'Cancelled')) = count(*))
```

Comment:
Also, it would be great to do something similar for notify children and marking the job task complete. Don't do those yet because my current PR for the unipod will conflict